ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,410 @@
+                QC.counterexample "lastCheck" $ x QC.=== y
+            _ ->
+                error "impossible! lastCheck response"
+
+    watcherEvents <- dumpEvents varEvents
+
+    pure
+      $ QC.monadicIO
+      $ QSM.prettyCommands prettySM hist
+      $ QC.counterexample
+            (unlines
+               $ (:) "WATCHER"
+               $ map (("    " <>) . show)
+               $ watcherEvents
+            )
+      $ QC.tabulate
+            "Notables"
+            (case Set.toList $ mNotables model' of
+               []       -> ["<none>"]
+               notables -> map show notables
+            )
+      $ QSM.checkCommandNames cmds
+      $ noExn QC..&&. lastCheck QC..&&. res QC.=== QSM.Ok
+
+-----
+
+durationUntilTooOld :: Selection -> IOSim.IOSim s GSM.DurationFromNow
+durationUntilTooOld sel = do
+    let expiryAge = ageLimit `SI.addTime` onset sel
+    now <- SI.getMonotonicTime
+    pure $ case compare expiryAge now of
+        LT -> GSM.Already
+        GT -> GSM.After $ realToFrac $ expiryAge `SI.diffTime` now
+
+        -- 'boringDur' cannot prevent this case. In particular, this case
+        -- necessarily arises in the GSM itself during a 'TimePasses' that
+        -- incurs a so-called /flicker/ event, in which the anti-thrashing
+        -- timer expires and yet the node state at that moment still
+        -- _immediately_ indicates that it's CaughtUp. For the specific case of
+        -- this test suite, the answer here must be 'GSM.Already'.
+        EQ -> GSM.Already
+
+-----
+
+-- | Ensure the GSM thread's transactions quiesce
+--
+-- I'm unsure how many 'yield's are actually necessary, but ten is both small
+-- and also seems likely to suffice.
+--
+-- Despite the crudeness, this seems much more compositional than invasive
+-- explicit synchronization.
+yieldSeveralTimes :: MonadFork m => m ()
+yieldSeveralTimes = replicateM_ 10 yield
+
+{-
+
+Note [Why yield after the command]
+
+For this 'prop_sequential1' repro
+
+@
+   YoungEnough
+
+   Command (NewCandidate Amara (B 1)) Unit []
+   Command (StartIdling Amara) Unit []
+   Command (TimePasses 61) Unit []
+   Command (ExtendSelection (S (-4))) Unit []
+   Command ReadMarker (ReadThisMarker Absent) []
+
+   (Command ReadJudgment _ [])   -- this last command is baked into the property
+@
+
+If we yield after the command, then both GSM flicker writes happen during the
+'ExtendSelection'.
+
+If we yield before the command, then both GSM flicker writes happen during the
+'ReadMarker'.
+
+If we don't yield, one write happens during the ReadMarker and the other
+happens /between/ 'ReadMarker' and 'ReadJudgment'.
+
+It seems most intuitive for the updates to happen "as part of" the
+'ExtendSelection', so I'm favoring yielding after.
+
+And since we're yielding after the command, we should also yield before the
+first command, for consistency.
+
+-}
+
+----- trivial event accumulator, useful for debugging test failures
```

Comment:
I tried this again, after our discussion. It still feels too much like ceremony. The root cause is that `recordTracerTVar` doesn't take the monotonic time. This gives me two options:

- Keep all the declarations I currently have, but change the implementation to use an underlying `Tracer`. This is a bit "overly complicated" since `recordTracerTVar` returns the read action separately from the tracer. It's two closures that share a hidden variable instead of the simple transparent first-order newtype in my current approach.
- Drop my definitions and replace them with `recordTracerTVar` and a similar combinator that adds in the montonic time. But at this point, `Vars` becomes a misnomer, since the `Tracer` is doing more than "just" wrapping a var: it's also sampling the time.

So---since I'm not otherwise using `Tracer`s---either options seems excessively disruptive.

Reply:
Also, the "few lines" I reuse are lost to imports and wrappers lol

Reply:
> since I'm not otherwise using `Tracer`s

This made me think: Wouldn't it be natural to have a tracer in the GSM logic when it does a state transition? This sounds very useful for debugging a full node. Then, the `EvGsmWrite` events could come from that tracer instead of a watcher.

Reply:
I added the tracer and removed the watcher 👍